pkg/activator/throttler.go

```diff
@@ -271,3 +288,52 @@ func (t *Throttler) endpointsDeleted(obj interface{}) {
 	revID := RevisionID{ep.Namespace, name}
 	t.Remove(revID)
 }
+
+type infiniteBreaker struct {
+	mu          sync.RWMutex
+	broadcast   chan struct{}
+	concurrency int
+}
+
+func (ib *infiniteBreaker) Capacity() int {
+	ib.mu.RLock()
+	defer ib.mu.RUnlock()
+	return ib.concurrency
+}
+
+func (ib *infiniteBreaker) UpdateConcurrency(cc int) error {
+	ib.mu.Lock()
+	defer ib.mu.Unlock()
+
+	old := ib.concurrency
+	ib.concurrency = cc
+	switch {
+	// Scaled to 0.
+	case cc == 0 && old != 0:
+		ib.broadcast = make(chan struct{})
+		// Scaled from 0.
+	case cc > 0 && old == 0:
+		close(ib.broadcast)
+		// No change.
+	case cc == 0 && old == 0 || cc > 0 && old > 0:
+		break
+	}
+	return nil
+}
+
+func (ib *infiniteBreaker) Maybe(ctx context.Context, thunk func()) bool {
+	has := ib.Capacity()
+	// We're scaled to serve.
+	if has > 0 {
+		thunk()
+		return true
+	}
+	select {
+	case <-ib.broadcast:
```

Comment:
For my own education: What happens if you try this on a closed channel? Will it return immediately or will it block?